src/mlpack/methods/ann/layer/log_softmax_impl.hpp

```diff
@@ -103,7 +103,7 @@ void LogSoftMaxType<MatType>::Backward(
     const MatType& gy,
     MatType& g)
 {
-  g = arma::exp(input) + gy;
+  g = gy - arma::exp(input) % arma::repmat(arma::sum(gy), input.n_rows, 1);
```

Comment:
A fascinating observation that was confusing me very much until I realized: these two implementations are the same when (a) `LogSoftMax` is the last layer in the network, and (b) `NegativeLogLikelihood` is the loss function used.

The reason is that with `NegativeLogLikelihood`, `gy` will always be such that only one element takes the value of `-1`, and all other values are `0`.  Thus, `-arma::exp(input) % arma::repmat(arma::sum(gy), input.rows, 1)` is equivalent to `-arma::exp(input) % -1`, which is just `arma::exp(input)`... and we get the original implementation.

So, when I tested this on the `mnist_simple` example using either implementation... the results were exactly the same!

I couldn't find any other examples in the examples repository that used `LogSoftMax` as the last layer but a different loss function than `NegativeLogLikelihood`.

In any case, after using @mrityunjay-tripathi's derivation in #2364, I agree that this is the correct implementation.

Reply:
That's definitely a interesting observation and also a relief knowing that all the models that uses this layer were not incorrectly trained!